### ISSUE TITLE
Target WSGI script '/opt/python/current/app/miclave/wsgi.py' cannot be loaded as Python module.

### ISSUE THREAD

**Issue by esilva-dt** [GrahamDumpleton/mod_wsgi#589]
Good evening.

I'm having this issue:
```
[Thu Jul 09 05:26:14.451547 2020] [:error] [pid 25499] [remote 172.105.89.161:72] mod_wsgi (pid=25499): Target WSGI script '/opt/python/current/app/miclave/wsgi.py' cannot be loaded as Python module.
[Thu Jul 09 05:26:14.451604 2020] [:error] [pid 25499] [remote 172.105.89.161:72] mod_wsgi (pid=25499): Exception occurred processing WSGI script '/opt/python/current/app/miclave/wsgi.py'.
[Thu Jul 09 05:26:14.451681 2020] [:error] [pid 25499] [remote 172.105.89.161:72] Traceback (most recent call last):
[Thu Jul 09 05:26:14.451703 2020] [:error] [pid 25499] [remote 172.105.89.161:72]   File "/opt/python/current/app/miclave/wsgi.py", line 12, in <module>
[Thu Jul 09 05:26:14.451708 2020] [:error] [pid 25499] [remote 172.105.89.161:72]     from django.core.wsgi import get_wsgi_application
[Thu Jul 09 05:26:14.451723 2020] [:error] [pid 25499] [remote 172.105.89.161:72] ModuleNotFoundError: No module named 'django'
```

I'm using AWS' elastic beanstalk, Django was installed in a virtual environment (/opt/python/run/venv/) during deployment as it was part of the requirements.txt file. I tried running the commands to get information related to Django and its installation and I got:

     /opt/python/run/venv/local/lib/python3.6/site-packages/django/__init__.py

I got two conf files in which mod_wsgi is invoked.

**A: ssl.conf**
```
LoadModule wsgi_module modules/mod_wsgi.so
WSGIPythonHome /opt/python/run/baselinenv
WSGISocketPrefix run/wsgi
WSGIRestrictEmbedded On
Listen 443
<VirtualHost *:443>
  SSLEngine on
  SSLCertificateFile "/etc/pki/tls/certs/server.crt"
  SSLCertificateKeyFile "/etc/pki/tls/certs/server.key"
  
  Alias /static/ /opt/python/current/app/static/
  <Directory /opt/python/current/app/static>
  Order allow,deny
  Allow from all
  </Directory>
  
  WSGIScriptAlias / /opt/python/current/app/miclave/wsgi.py
  
  <Directory /opt/python/current/app>
  Require all granted
  </Directory>
  
  WSGIDaemonProcess wsgi-ssl processes=1 threads=15 display-name=%{GROUP} \
    python-path=/opt/python/current/app:/opt/python/run/venv/lib64/python3.6/site-packages:/opt/python/run/venv/lib/python3.6/site-packages \
    home=/opt/python/current/app \
    user=wsgi \
    group=wsgi
  WSGIProcessGroup wsgi-ssl
  
</VirtualHost>
```

**B: wsgi.conf**
```
LoadModule wsgi_module modules/mod_wsgi.so
WSGIPythonHome /opt/python/run/baselinenv
WSGISocketPrefix run/wsgi
WSGIRestrictEmbedded On

<VirtualHost *:80>

Alias /static/ /opt/python/current/app/static/
<Directory /opt/python/current/app/static/>
Order allow,deny
Allow from all
</Directory>


WSGIScriptAlias / /opt/python/current/app/miclave/wsgi.py


<Directory /opt/python/current/app/>
  Require all granted
</Directory>

WSGIDaemonProcess wsgi processes=1 threads=15 display-name=%{GROUP} \
  python-home=/opt/python/run/venv/ \
  python-path=/opt/python/current/app user=wsgi group=wsgi \
  home=/opt/python/current/app
WSGIProcessGroup wsgi
</VirtualHost>

LogFormat "%h (%{X-Forwarded-For}i) %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\"" combined
```

I tried adding the python-home option in the WSGIDaemonProcess in **ssl.conf** but I got another error.

I'm at a loss here, I would appreciate any recommendation. Thanks beforehand.

**Comment by GrahamDumpleton** [comment]
Quick question.

Did you remove ``ServerName`` directive from what you posted just so not exposing host name? Usually you would have a ``ServerName`` directive, especially since wanting to run a secure site.

I don't need to see what you have for ``ServerName`` and ``ServerAlias`` if set, but just want to know if they exist.

**Comment by esilva-dt** [comment]
The directive exists in the **httpd.conf** file. The **wsgi.py** file is generated automatically by AWS while the **ssl.conf** file is created after adding new directives to allow https.

**Comment by GrahamDumpleton** [comment]
AWS always screws things up when they auto generate stuff for mod_wsgi and don't follow goo practices. :-(

Anyway, ideally the following should only appear once, not in multiple places.

```
LoadModule wsgi_module modules/mod_wsgi.so
WSGISocketPrefix run/wsgi
WSGIRestrictEmbedded On
```

You don't need ``WSGIPythonHome`` or a base line virtual environment since using daemon mode exclusively.

The two ``VirtualHost`` should then be:

```
<VirtualHost *:443>
  SSLEngine on
  SSLCertificateFile "/etc/pki/tls/certs/server.crt"
  SSLCertificateKeyFile "/etc/pki/tls/certs/server.key"
  
  Alias /static/ /opt/python/current/app/static/

  <Directory /opt/python/current/app/static>
  Order allow,deny
  Allow from all
  </Directory>
  
  WSGIScriptAlias / /opt/python/current/app/miclave/wsgi.py
  
  <Directory /opt/python/current/app>
  Require all granted
  </Directory>
  
  WSGIDaemonProcess wsgi-ssl processes=1 threads=15 display-name=%{GROUP} \
    python-home=/opt/python/run/venv/local \
    python-path=/opt/python/current/app \
    home=/opt/python/current/app \
    user=wsgi \
    group=wsgi

  WSGIProcessGroup wsgi-ssl
  WSGIApplicationGroup %{GLOBAL}
  
</VirtualHost>

<VirtualHost *:80>
Alias /static/ /opt/python/current/app/static/

<Directory /opt/python/current/app/static/>
Order allow,deny
Allow from all
</Directory>

WSGIScriptAlias / /opt/python/current/app/miclave/wsgi.py

<Directory /opt/python/current/app/>
  Require all granted
</Directory>

WSGIProcessGroup wsgi-ssl
WSGIApplicationGroup %{GLOBAL}
</VirtualHost>
```

In other words, the ``python-home`` was missing on the 443 ``VirtualHost`` and was wrong value on 80 ``VirtualHost``.

You don't even need the ``WSGIDaemonProcess`` in 80 ``VirtualHost``, as ``WSGIProcessGroup`` can refer across to one in 443 ``VirtualHost``.

Both ``VirtualHost`` were missing the recommend ``WSGIApplicationGroup`` directive.

**Comment by esilva-dt** [comment]
I updated my conf files as you explained and it worked smoothly! Many thanks, Mr. Dumpleton, you're a real hero!

> AWS always screws things up when they auto generate stuff for mod_wsgi and don't follow goo practices. :-(
You're absolutely right, there were many things missing and some others were redundant. But now I got this correct base so I'll use it whenever I face a similar situation, hopefully,  it won't be anytime soon.

Thanks again for your invaluable help! Hope you have a nice day and take care out there!
